### PR TITLE
Convert old first edition links to current edition one

### PR DESCRIPTION
**Pull request by tesuji**
r? @steveklabnik

### CODE REVIEW

**Comment by frewsxcv** [comment]
Tidy checker wants some lines to be wrapped: https://travis-ci.com/rust-lang/rust/jobs/172317538

File: src/librustc/diagnostics.rs [hunk 1]
```diff
@@ -642,7 +642,7 @@ attributes:
 #![no_std]
 ```
 
-See also https://doc.rust-lang.org/book/first-edition/no-stdlib.html
+See also https://doc.rust-lang.org/unstable-book/language-features/lang-items.html#writing-an-executable-without-stdlib
 "##,
```
Comment (lzutao) [inline src/librustc/diagnostics.rs]: I just don't know how to wrap this line.

**Comment by Dylan-DPC** [comment]
ping from triage @steveklabnik awaiting your review on this

**Comment by steveklabnik** [comment]
This looks good to me, but I don't know how to wrap that line.

r=me once CI gets sorted out, but I'm not sure who to ask. Maybe @rust-lang/infra ?

File: src/librustc/diagnostics.rs (hunk 1 above)
Comment (kennytm) [reply src/librustc/diagnostics.rs]: You could reword it like 

```markdown
See also [«some description here»][1].

[1]: https://doc.rust-lang.org/unstable-book/language-features/lang-items.html#writing-an-executable-without-stdlib
```

Comment (lzutao) [reply src/librustc/diagnostics.rs]: The link itself is longer than 100 characters, which is prohibited by the lint. :man_shrugging:

Comment (kennytm) [reply src/librustc/diagnostics.rs]: @lzutao Tidy allows long lines written in the form `[stuff]: http://url`.

https://github.com/rust-lang/rust/blob/a54b5c7a645ead203d77e78245362f9e0f00dd3c/src/tools/tidy/src/style.rs#L47-L53

**Comment by kennytm** [comment]
@lzutao @steveklabnik 

I think we could simply add `// ignore-tidy-linelength` to the file, after applying https://github.com/rust-lang/rust/pull/57856#discussion_r255814002.

The `register_long_diagnostics!` macro will eventually invoke the `__register_diagnostic!` syntax extension. There is already a length check in it.

https://github.com/rust-lang/rust/blob/a54b5c7a645ead203d77e78245362f9e0f00dd3c/src/libsyntax/diagnostics/plugin.rs#L101-L107

So we don't need tidy to double-check it.

Alternative: Instead of `// ignore-tidy-linelength` we could relax tidy's [`line_is_url()` function](https://github.com/rust-lang/rust/blob/a54b5c7a645ead203d77e78245362f9e0f00dd3c/src/tools/tidy/src/style.rs#L53) to accept `[stuff]: http://url` without being a comment when the file name is `diagnostic.rs`, but this change is much more involved.

File: src/librustc/diagnostics.rs [hunk 2]
```diff
@@ -406,7 +407,7 @@ fn baz<'a>(x: &'a str, y: &str) -> &str { }
 Lifetime elision in implementation headers was part of the lifetime elision
 RFC. It is, however, [currently unimplemented][iss15872].
```
Comment (lzutao) [inline src/librustc/diagnostics.rs]: The issue  #15872 is closed so this statement is no longer true. What should we do with this error code?

File: src/librustc_typeck/diagnostics.rs [hunk 3]
```diff
@@ -3024,19 +3028,20 @@ fn coerce_foo<T: CoerceUnsized<U>, U>(t: T) -> Foo<U> {
...
 whereas a struct containing an unsized type needs to know the unsized type
-field it's containing is able to be coerced. An
-[unsized type](https://doc.rust-lang.org/book/first-edition/unsized-types.html)
+field it's containing is able to be coerced. An [unsized type][1]
 is any type that the compiler doesn't know the length or alignment of at
 compile time. Any struct containing an unsized type is also unsized.
 
+[1]: https://doc.rust-lang.org/book/ch19-04-advanced-types.html#dynamically-sized-types-and-the-sized-trait
+
 Example of erroneous code:
```
Comment (lzutao) [inline src/librustc_typeck/diagnostics.rs]: Need a review about this style. cc @steveklabnik

File: src/librustc/diagnostics.rs (hunk 2 above)
Comment (steveklabnik) [reply src/librustc/diagnostics.rs]: IMHO, this whole paragraph should just be removed.

File: src/librustc_typeck/diagnostics.rs (hunk 3 above)
Comment (steveklabnik) [reply src/librustc_typeck/diagnostics.rs]: looks great!

File: src/librustc/diagnostics.rs (hunk 2 above)
Comment (lzutao) [reply src/librustc/diagnostics.rs]: I will open another PR to fix that after this one is merged.

**Comment by lzutao** [comment]
Finally travis is green. Should I squash it into single commit?

File: src/libcore/convert.rs [hunk 4]
```diff
@@ -113,9 +113,6 @@ pub const fn identity<T>(x: T) -> T { x }
 /// - Use `Borrow` when the goal is related to writing code that is agnostic to
 ///   the type of borrow and whether it is a reference or value
 ///
-/// See [the book][book] for a more detailed comparison.
-///
-/// [book]: ../../book/first-edition/borrow-and-asref.html
 /// [`Borrow`]: ../../std/borrow/trait.Borrow.html
```
Comment (lzutao) [inline src/libcore/convert.rs]: I cannot find any equivalent link to this comparison in the new book.

Comment (lzutao) [reply src/libcore/convert.rs]: cc @steveklabnik

Comment (steveklabnik) [reply src/libcore/convert.rs]: That’s correct, there’s not, which is why removing it is the right call.

**Comment by steveklabnik** [comment]
Yeah a squash would be great! r=me after

File: src/librustc/diagnostics.rs (hunk 2 above)
Comment (lzutao) [reply src/librustc/diagnostics.rs]: I opened a PR #58456

**Pull request merged**